File: src/main/java/com/camackenzie/exvi/core/model/ExerciseSet.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.camackenzie.exvi.core.model;

/**
 *
 * @author callum
 */
public class ExerciseSet {

    private final Exercise exercise;
    private int[] sets;
    private String unit;

    public ExerciseSet(Exercise ex, String unit, int... sets) {
        this.exercise = ex;
        this.sets = sets;
    }

    public void setSet(int index, int count) {
        this.sets[index] = count;
    }

    public int getSet(int index) {
        return this.sets[index];
    }

    public Exercise getExercise() {
        return this.exercise;
    }

    public int[] getSets() {
        return this.sets;
    }

    public void setSets(int... sets) {
        this.sets = sets;
    }

    public String getUnit() {
        return this.unit;
    }

    public void setUnit(String unit) {
        this.unit = unit;
    }

    public static ExerciseSet repSets(Exercise ex, int... sets) {
        return new ExerciseSet(ex, "rep", sets);
    }

    public static ExerciseSet secondSets(Exercise ex, int... sets) {
        return new ExerciseSet(ex, "second", sets);
    }

    public static ExerciseSet minuteSets(Exercise ex, int... sets) {
        return new ExerciseSet(ex, "minute", sets);
    }

}

File: exvi-core/src/commonMain/kotlin/com/camackenzie/exvi/core/model/ExerciseSet.kt
/*
 * Copyright (c) Callum Mackenzie 2022.
 */
package com.camackenzie.exvi.core.model

import com.camackenzie.exvi.core.util.SelfSerializable
import kotlinx.serialization.Polymorphic
import kotlin.jvm.JvmStatic

@Suppress("unused")
interface ExerciseSet : SelfSerializable {
    @Polymorphic
    var exercise: Exercise
    var unit: String

    @Polymorphic
    val sets: MutableList<SingleExerciseSet>

    operator fun component1(): Exercise = exercise
    operator fun component2(): String = unit
    operator fun component3(): MutableList<SingleExerciseSet> = sets

    // Converts the inner exercise to standard of same name if present
    // Returns whether the exercise was standardized or not
    fun tryStandardize(): Boolean {
        val standard = exercise.tryStandardize()
        return if (standard != null) {
            exercise = standard
            true
        } else false
    }

    fun toActual() = ActualExerciseSet(exercise, unit, ArrayList(sets.map {
        it.toActual()
    }))

    companion object {
        /**
         * Constructs a new ActualExerciseSet object
         */
        @JvmStatic
        operator fun invoke(
            exercise: Exercise,
            unit: String,
            sets: List<SingleExerciseSet>
        ) = ActualExerciseSet(exercise, unit, ArrayList(sets))

        /**
         * Constructs a new ActualExerciseSet object
         */
        @JvmStatic
        operator fun invoke(
            exercise: Exercise, unit: String, sets: Array<Int>
        ) = invoke(exercise, unit, sets.map { SingleExerciseSet(it) }.toList())

        /**
         * Constructs a new ActualExerciseSet object
         */
        @JvmStatic
        fun repSets(ex: Exercise, sets: List<SingleExerciseSet>) = ExerciseSet(ex, "rep", sets)

        /**
         * Constructs a new ActualExerciseSet object
         */
        @JvmStatic
        fun secondSets(ex: Exercise, sets: List<SingleExerciseSet>) = ExerciseSet(ex, "second", sets)

        /**
         * Constructs a new ActualExerciseSet object
         */
        @JvmStatic
        fun minuteSets(ex: Exercise, sets: List<SingleExerciseSet>) = ExerciseSet(ex, "minute", sets)
    }
}

